packages/components/src/TooltipTrigger/TooltipTrigger.component.js

```diff
@@ -24,124 +35,168 @@ const props = {
-
-	/**
-	 * Show the tooltip after mouse up
-	 */
-	onMouseUp = (...args) => {
-		this.overlay.handleDelayedShow();
-		if (this.props.children.props.onMouseUp) {
-			this.props.children.props.onMouseUp(...args);
-		}
-	};
-
-	/**
-	 * when activate an element, hide the tooltip between keydown & keyup
-	 * @param {object} event the keyDown event
-	 */
-	onKeyDown = event => {
-		if (event.which === keycode.codes.enter || event.which === keycode.codes.space) {
-			this.overlay.handleDelayedHide();
+	componentWillUnmount() {
+		clearTimeout(this.timeout);
+	}
+
+	getAdjustedTooltipPlacement(tooltipPlacement, dimensions) {
+		const { tooltipHeight = DEFAULT_OFFSET_Y, tooltipWidth = DEFAULT_OFFSET_X } = this.props;
+		let placement = tooltipPlacement;
+		if (dimensions.bottom + tooltipHeight > window.innerHeight) {
+			if (['left', 'right'].includes(tooltipPlacement)) {
+				placement = `${tooltipPlacement}-bottom`;
+			} else if (tooltipPlacement === 'bottom') {
+				placement = 'top';
+			}
+		} else if (dimensions.top - tooltipHeight < 0) {
+			if (tooltipPlacement === 'top') {
+				placement = 'bottom';
+			}
 		}
-	};
-
-	/**
-	 * when activate an element, hide the tooltip between keydown & keyup
-	 * @param {object} event the keyup event
-	 */
-	onKeyUp = event => {
-		if (event.which === keycode.codes.enter || event.which === keycode.codes.space) {
-			this.overlay.handleDelayedShow();
+		if (dimensions.left - tooltipWidth / 2 < 0) {
+			if (['top', 'bottom'].includes(tooltipPlacement)) {
+				placement = `${placement}-right`;
+			} else if (tooltipPlacement === 'left') {
+				placement = placement.replace('left', 'right');
+			}
+		} else if (dimensions.right + tooltipWidth / 2 > window.innerWidth) {
+			if (['top', 'bottom'].includes(tooltipPlacement)) {
+				placement = `${placement}-left`;
+			} else if (tooltipPlacement === 'right') {
+				placement = placement.replace('right', 'left');
+			}
 		}
-	};
+		return placement;
+	}
 
-	render() {
-		const child = React.Children.only(this.props.children);
+	getTooltipPosition() {
+		const { tooltipPlacement = 'right', tooltipWidth = DEFAULT_OFFSET_X } = this.props;
 
-		if (!this.state.hovered) {
-			return cloneElement(child, {
-				onMouseOver: this.onMouseOver,
-				onFocus: this.onFocus,
-			});
+		if (!this.el) {
+			return {
+				tooltipPlacement,
+			};
 		}
 
-		const tooltip = (
-			<Tooltip className={getTooltipClass()} id={this.state.id}>
-				{this.props.label}
-			</Tooltip>
-		);
-		// TODO jmfrancois : render the Tooltip in a provider so use context for that.
-		// fix with disabling element https://github.com/react-bootstrap/react-bootstrap/pull/3251
-		// we add onClick={() => this.overlay.handleDelayedHide() to hide the overlay
-		// it stays when the element inside is disabled
+		const dimensions = this.el.getBoundingClientRect();
+
+		const placement = this.getAdjustedTooltipPlacement(tooltipPlacement, dimensions);
+
+		return {
+			placement,
+			style: {
+				top: (() => {
```

Comment:
Same as above, this is a big if forest...

Reply:
I will do my best